File: api/src/controllers/upload.controller.ts
import { Request, Response } from 'express';
import { DocumentService } from '../services/document.service';
import { EmbeddingService } from '../services/embedding.service';
import { DatabaseModel } from '../models/database.model';
import { MinIOModel } from '../models/minio.model';


export class UploadController {
  constructor(
    private db: DatabaseModel,
    private minio: MinIOModel,
    private documentService: DocumentService,
    private embeddingService: EmbeddingService
  ) { }

  // INGEST từ MinIO → PostgreSQL
  async ingestFromMinIO(req: Request, res: Response) {
    try {
      const { objectName } = req.body;
      if (!objectName) {
        return res.status(400).json({ error: 'Thiếu objectName' });
      }

      // Lấy file từ MinIO
      const buffer = await this.minio.getFile(objectName);

      // Extract + clean text
      const rawText = await this.documentService.extractText(buffer, objectName);
      const text = this.documentService.cleanText(rawText);

      if (!text) {
        return res.status(400).json({ error: 'File không có nội dung text' });
      }

      const metadata = this.documentService.parseMetadataFromPath(objectName, text);

      // Insert document → lấy document_id
      const documentId = await this.db.insertDocument({
        filename: objectName,
        file_path: objectName,
        file_size: buffer.length,
        content_type: 'unknown',
        document_type: metadata.document_type,
        metadata: metadata as any,
      });

      //  Chunk text - đoạn này HA sửa theo metadata từng chunk đã thiết lập mới
      const chunks = this.documentService.chunkText(text);


      // Embed + insert chunks - HA cũng sửa vòng for
      for (let i = 0; i < chunks.length; i++) {
        const chunkText = chunks[i];
        if (!chunkText) continue;

        const embedding = await this.embeddingService.generateEmbedding(chunkText);

        if (embedding.length !== 1536) {
          throw new Error(`Embedding dimension invalid: ${embedding.length}`);
        }

        await this.db.insertChunk({
          document_id: documentId,
          content: chunkText,
          chunk_index: i,
          embedding,
        });
      }


      return res.json({
        message: 'Ingest thành công',
        documentId,
        totalChunks: chunks.length,
      });
    } catch (error: any) {
      console.error('Ingest error:', error);
      return res.status(500).json({ error: error.message });
    }
  }

  // List files trong MinIO
  async listMinIOFiles(req: Request, res: Response) {
    try {
      const prefix = (req.query.prefix as string) || '';
      const files = await this.minio.listFiles(prefix);
      return res.json({
        files,
        total: files.length,
      });
    } catch (error: any) {
      console.error('List files error:', error);
      return res.status(500).json({ error: error.message });
    }
  }

  // Kiểm tra trạng thái upload
  async checkUploadStatus(req: Request, res: Response) {
    try {
      const stats = await this.db.getIngestStats();

      return res.json({
        summary: {
          totalDocuments: Number(stats.totalDocuments || 0),
          totalChunks: Number(stats.totalChunks || 0),
          totalSize: 0,
        },
        documents: stats.documents || [],
      });
    } catch (error: any) {
      console.error('Check status failed:', error);
      res.status(500).json({ error: error.message });
    }
  }
}

File: api/backend/controllers/upload_controller.py
from __future__ import annotations

import os
from typing import Any

from backend.models.database import DatabaseModel
from backend.models.minio_model import MinIOModel
from backend.services.document_service import DocumentService
from backend.services.embedding_service import EmbeddingService


class UploadController:
    def __init__(
        self,
        *,
        db: DatabaseModel,
        minio: MinIOModel,
        document_service: DocumentService,
        embedding_service: EmbeddingService,
    ) -> None:
        self.db = db
        self.minio = minio
        self.document_service = document_service
        self.embedding_service = embedding_service

    def ingest_from_minio(self, payload: dict[str, Any]) -> tuple[dict[str, Any], int]:
        object_name = payload.get("objectName")
        if not object_name:
            return {"error": "Missing objectName"}, 400
        ingest_batch_size = max(1, int(os.getenv("INGEST_BATCH_SIZE") or 32))

        try:
            file_buffer = self.minio.get_file(str(object_name))
            raw_text = self.document_service.extract_text(file_buffer, str(object_name))
            text = self.document_service.clean_text(raw_text)

            if not text:
                return {"error": "No extractable text in file"}, 400

            metadata = self.document_service.parse_metadata_from_path(str(object_name), text)

            document_id = self.db.insert_document(
                filename=str(object_name),
                file_path=str(object_name),
                file_size=len(file_buffer),
                content_type="unknown",
                document_type=str(metadata.get("document_type") or "syllabus"),
                metadata=metadata,
            )

            chunks = [chunk for chunk in self.document_service.chunk_text(text) if chunk]
            for start in range(0, len(chunks), ingest_batch_size):
                chunk_batch = chunks[start : start + ingest_batch_size]
                embedding_batch = self.embedding_service.generate_batch_embeddings(chunk_batch)
                self.db.insert_chunks_batch(
                    document_id=document_id,
                    chunks=chunk_batch,
                    embeddings=embedding_batch,
                    start_index=start,
                )

            return {
                "message": "Ingest completed",
                "documentId": document_id,
                "totalChunks": len(chunks),
            }, 200
        except Exception as exc:
            return {"error": str(exc)}, 500

    def list_minio_files(self, prefix: str = "") -> tuple[dict[str, Any], int]:
        try:
            files = self.minio.list_files(prefix)
            return {"files": files, "total": len(files)}, 200
        except Exception as exc:
            return {"error": str(exc)}, 500

    def check_upload_status(self) -> tuple[dict[str, Any], int]:
        try:
            stats = self.db.get_ingest_stats()
            return {
                "summary": {
                    "totalDocuments": int(stats.get("totalDocuments") or 0),
                    "totalChunks": int(stats.get("totalChunks") or 0),
                    "totalSize": 0,
                },
                "documents": stats.get("documents") or [],
            }, 200
        except Exception as exc:
            return {"error": str(exc)}, 500
